### PR TITLE
P2p framework with discover.

### PR DESCRIPTION
**Pull request by qiaozhigang**

### CODE REVIEW

File: p2p/discovery/message.go [hunk 1]
```diff
@@ -119,7 +119,7 @@ func (m *findNode) handle(t *udp, from *net.UDPAddr) {
 	log.Debug("received find node request from: %s", common.BytesToHex(m.SelfID.Bytes()))
 
 	node := NewNodeWithAddr(m.SelfID, from)
-	t.table.addNode(node)
+	t.table.AddNode(node)
```
Comment (lucusfly) [inline p2p/discovery/message.go]: keep this one, use udp.addNode

File: p2p/discovery/server.go [hunk 2]
```diff
@@ -27,6 +27,18 @@ func StartServer(port, id string) {
...
 	wg.Add(1)
 	wg.Wait()
+
+}
+
+func StartServerTest(port, id string, nodeArr []*Node) (db *Database) {
+	udp := getUDP(port, HexToNodeID(id))
+	log.Debug("nodeid: %s", common.BytesToHex(udp.self.ID.Bytes()))
+	for _, node := range nodeArr {
+		udp.table.AddNode(node)
```
Comment (lucusfly) [inline p2p/discovery/server.go]: use udp.addNode()

File: p2p/message.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+	size       uint32 // size of the paylod
+	payload    []byte
+	ReceivedAt time.Time
+	CurPeer    *Peer
+}
+
+// msg
+type msg struct {
+	Message
+	protoCode uint16
```
Comment (lucusfly) [inline p2p/message.go]: what's the different between msg and Message

File: p2p/peer.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/seeleteam/go-seele/p2p/discovery"
+)
+
+type Peer struct {
+	fd       net.Conn
```
Comment (lucusfly) [inline p2p/peer.go]: add comments for fd

File: p2p/test/test.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
+package main
+
```
Comment (lucusfly) [inline p2p/test/test.go]: remove this file

File: p2p/server.go [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+	kadDB    *discovery.Database
+	listener net.Listener
+
+	//ourHandshake *protoHandshake
+	//lastLookup time.Time
+
+	// These are for Peers, PeerCount (and nothing else).
+	//peerOp     chan peerOpFunc
+	//peerOpDone chan struct{}
+
```
Comment (lucusfly) [inline p2p/server.go]: remove unused code

File: p2p/server.go [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+		}
+	}
+
+	//Trace("P2P networking is spinning down")
+
+	// Terminate discovery. If there is a running lookup it will terminate soon.
+	/*if srv.ntab != nil {
+		srv.ntab.Close()
+	}*/
+
```
Comment (lucusfly) [inline p2p/server.go]: please clear unused code

File: p2p/server.go [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+	checkTimer := time.NewTimer(3 * time.Second)
+running:
+	for {
+		srv.scheduleTasks()
+		select {
+		case <-checkTimer.C:
+			checkTimer.Reset(3 * time.Second)
+		case <-srv.quit:
+			// The server was stopped. Run the cleanup logic.
+			break running
```
Comment (lucusfly) [inline p2p/server.go]: what's the difference with continue

File: p2p/server.go [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+		case c := <-srv.addpeer:
+			fmt.Println("<-srv.addpeer", c)
```
Comment (lucusfly) [inline p2p/server.go]: use log.Debug() in case you forgot delete it

File: p2p/server.go [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+	// These are for Peers, PeerCount (and nothing else).
+	//peerOp     chan peerOpFunc
+	//peerOpDone chan struct{}
+
+	quit chan struct{}
+	//addstatic     chan *discover.Node
+	//removestatic  chan *discover.Node
+	//posthandshake chan *conn
+	addpeer chan *Peer
+	delpeer chan *Peer
```
Comment (lucusfly) [inline p2p/server.go]: delete peer?

File: p2p/server.go [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+		//Trace("<-delpeer (spindown)", "remainingTasks", len(runningTasks))
+		delete(peers, p.node.ID)
+	}
+}
+
+//scheduleTasks
+func (srv *Server) scheduleTasks() {
+	// TODO select nodes from ntab to connect
+	nodeMap := srv.kadDB.GetCopy()
+	fmt.Println("scheduleTasks called...", nodeMap)
```
Comment (lucusfly) [inline p2p/server.go]: use log.Debug()

File: p2p/server.go [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+		go func() {
+			srv.setupConn(fd, inboundConn, nil)
+			slots <- struct{}{}
+		}()
+	}
+}
+
+// setupConn todo handshake.
+func (srv *Server) setupConn(fd net.Conn, flags int, dialDest *discovery.Node) error {
+	peer := &Peer{
```
Comment (lucusfly) [inline p2p/server.go]: too long method, refine it

File: p2p/server.go [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+	for _, proto := range srv.Protocols {
+		caps = append(caps, proto.GetBaseProtocol().cap())
+	}
+	hsMsg := &msg{
+		protoCode: ctlProtoCode,
+		Message: Message{
+			msgCode: ctlMsgProtoHandshake,
+		},
+	}
+	//buffer := new(bytes.Buffer)
```
Comment (lucusfly) [inline p2p/server.go]: remove unused code

File: vendor/vendor.json [hunk 14]
```diff
@@ -4,233 +4,231 @@
 	"package": [
 		{
 			"checksumSHA1": "YdWmkLgfIqoZgQA0bWB71jCizpc=",
+			"origin": "github.com/seeleteam/vendor/github.com/aristanetworks/goarista/monotime",
 			"path": "github.com/aristanetworks/goarista/monotime",
-			"revision": "bdff49616af24181f3b83535ed73127c615ba55a",
-			"revisionTime": "2018-01-24T01:13:10Z"
```
Comment (lucusfly) [inline vendor/vendor.json]: why you delete them?

**Comment by herove** [comment]
@seeleteam/seele-dev Please join

File: p2p/discovery/server.go [hunk 15]
```diff
@@ -27,6 +27,18 @@ func StartServer(port, id string) {
 	wg := sync.WaitGroup{}
 	wg.Add(1)
 	wg.Wait()
+
```
Comment (WTer) [inline p2p/discovery/server.go]: remove this unnecessary blank line.

File: p2p/discovery/server.go [hunk 16]
```diff
@@ -27,6 +27,18 @@ func StartServer(port, id string) {
...
+func StartServerTest(port, id string, nodeArr []*Node) (db *Database) {
+	udp := getUDP(port, HexToNodeID(id))
+	log.Debug("nodeid: %s", common.BytesToHex(udp.self.ID.Bytes()))
+	for _, node := range nodeArr {
+		udp.table.AddNode(node)
+	}
+
+	udp.StartServe()
+	return udp.db
```
Comment (WTer) [inline p2p/discovery/server.go]: just discussion: 
generally, I suppose udp.StartServe() will be infinite service loop and never stop unless exit,
however you return udp.db, so I think this func name would be renamed to something like StartServerAndGetDB.

File: p2p/message.go [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+package p2p
+
+import (
+	"time"
+)
+
+// Message
+type Message struct {
+	msgCode    uint16
+	size       uint32 // size of the paylod
```
Comment (WTer) [inline p2p/message.go]: typo: paylod -> payload

File: p2p/peer.go [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+	go p.readLoop(readErr)
+	go p.pingLoop()
+
+	// Wait for an error or disconnect.
+loop:
+	for {
+		select {
+		case err = <-writeErr:
+			// A write finished. Allow the next write to start if
+			// there was no error.
```
Comment (WTer) [inline p2p/peer.go]: nice to refine the comments:
When a write operation finished without any error, the next write operation will continue.

File: p2p/peer.go [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+		//case err = <-p.disc:
+		case <-p.disc:
+			break loop
+		}
+	}
+
+	close(p.closed)
+	p.fd.Close()
+	p.wg.Wait()
+	// del peer from protocols
```
Comment (WTer) [inline p2p/peer.go]: del -> delete

we'd better make our comments more official.

File: p2p/peer.go [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+		return errors.New("not valid protoCode")
+	}
+	// for control msg
+	switch {
+	case msgRecv.msgCode == ctlMsgPingCode:
+		go p.sendCtlMsg(ctlMsgPongCode)
+	case msgRecv.msgCode == ctlMsgDiscCode:
+		//var reason [1]DiscReason
+
+		//rlp.Decode(msg.Payload, &reason)
```
Comment (WTer) [inline p2p/peer.go]: you can just remove lines #128 ~130 to make code more clean.

File: p2p/peer.go [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+		return errors.New("Not Found protoCode")
+	}
+	msgRaw := &msg{
+		protoCode: protoCode,
+		Message:   *msgSend,
+	}
+	return p.sendRawMsg(msgRaw)
+}
+
+func (p *Peer) sendCtlMsg(msgCode uint16) error {
```
Comment (WTer) [inline p2p/peer.go]: nice to rename this func name to sendControlMsg as Ctl is little meaningful for someone.

File: p2p/peer.go [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+	}
+	hsMsg.size = 0
+	p.sendRawMsg(hsMsg)
+	return nil
+}
+
+func (p *Peer) sendRawMsg(msgSend *msg) error {
+	p.wMutex.Lock()
+	defer p.wMutex.Unlock()
+	b := make([]byte, 8)
```
Comment (WTer) [inline p2p/peer.go]: nice to add a new blank line between #163 and 164

File: p2p/peer.go [hunk 23]
```diff
@@ -0,0 +1,218 @@
...
+	_, err := p.fd.Write(b)
+	//fmt.Println("sendRawMsg,head", sendLen)
+	if err != nil {
+		return err
+	}
+	_, err = p.fd.Write(msgSend.payload)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("sendRawMsg protoCode:%d msgCode:%d \r\n", msgSend.protoCode, msgSend.msgCode)
```
Comment (WTer) [inline p2p/peer.go]: use log.Info instead.

Comment (WTer) [reply p2p/peer.go]: also for other places.

File: p2p/protocol.go [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+	baseProtoCode uint   = 8
+	ctlProtoCode  uint16 = 1
+
+	ctlMsgProtoHandshake uint16 = 10
+	ctlMsgDiscCode       uint16 = 4
+	ctlMsgPingCode       uint16 = 3
+	ctlMsgPongCode       uint16 = 4
+)
+
+//Protocol base class
```
Comment (WTer) [inline p2p/protocol.go]: add a blank before Protocol

File: p2p/protocol.go [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+type Protocol struct {
+	// Protocol represents a P2P subprotocol implementation.
+	// Name should contain the official protocol name,
+	// often a three-letter word.
+	Name string //see dwn gss
```
Comment (WTer) [inline p2p/protocol.go]: dwn gss?

File: p2p/protocol.go [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+	go p.Run()
+	return
+}
+*/
+func (p *Protocol) cap() Cap {
+	return Cap{p.Name, p.Version}
+}
+
+// Cap is the structure of a peer capability.
+type Cap struct {
```
Comment (WTer) [inline p2p/protocol.go]: there are only 2 simple members in this struct, how can I understand the peer capability?

File: p2p/protocol.go [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+	Name    string
+	Version uint
+}
+
+/*
+// RlpData for RlpData
+func (cap Cap) RlpData() interface{} {
+	return []interface{}{cap.Name, cap.Version}
+}
+*/
```
Comment (WTer) [inline p2p/protocol.go]: you'd better to remove these commented code.

File: p2p/protocol.go [hunk 28]
```diff
@@ -0,0 +1,77 @@
...
+func (cap Cap) String() string {
+	return fmt.Sprintf("%s/%d", cap.Name, cap.Version)
+}
+
+type capsByNameAndVersion []Cap
+
+func (cs capsByNameAndVersion) Len() int      { return len(cs) }
+func (cs capsByNameAndVersion) Swap(i, j int) { cs[i], cs[j] = cs[j], cs[i] }
```
Comment (WTer) [inline p2p/protocol.go]: coding style: one line function seems not align with the common style

File: p2p/server.go [hunk 29]
```diff
@@ -0,0 +1,375 @@
...
+	}*/
+
+	srv.quit = make(chan struct{})
+	srv.addpeer = make(chan *Peer)
+	srv.delpeer = make(chan *Peer)
+	//srv.posthandshake = make(chan *conn)
+	//srv.addstatic = make(chan *discover.Node)
+	//	srv.removestatic = make(chan *discover.Node)
+	//	srv.peerOp = make(chan peerOpFunc)
+	//srv.peerOpDone = make(chan struct{})
```
Comment (WTer) [inline p2p/server.go]: remove these unused code.
also for others.

File: p2p/server.go [hunk 30]
```diff
@@ -0,0 +1,375 @@
...
+			checkTimer.Reset(3 * time.Second)
+		case <-srv.quit:
+			// The server was stopped. Run the cleanup logic.
+			break running
+		case c := <-srv.addpeer:
+			fmt.Println("<-srv.addpeer", c)
+			_, ok := peers[c.node.ID]
+			if ok {
+				// already connected
+				c.Disconnect(10)
```
Comment (WTer) [inline p2p/server.go]: nice to define this magic number as a meaningful constant

File: p2p/peer.go [hunk 31]
```diff
@@ -0,0 +1,222 @@
...
+	p.wg.Wait()
+	// send delpeer message for each protocols
+	for _, proto := range p.protoMap {
+		proto.DelPeerCh <- p
+	}
+	p.myLog.Infof("p2p.peer.run quit. err=%s", p.err)
+}
+
+func (p *Peer) pingLoop() {
+	ping := time.NewTimer(pingInterval)
```
Comment (lucusfly) [inline p2p/peer.go]: why you need ping loop, I think it's in discovery part

**Pull request closed without merging**